pl_examples/full_examples/imagenet/imagenet_example.py

```diff
@@ -0,0 +1,252 @@
+            'log': tqdm_dict
+        })
+
+        return output
+
+    def validation_step(self, batch, batch_idx):
+        images, target = batch
+        output = self.model(images)
+        loss_val = F.cross_entropy(output, target)
+        acc1, acc5 = self.__accuracy(output, target, topk=(1, 5))
+
+        # in DP mode (default) make sure if result is scalar, there's another dim in the beginning
+        if self.trainer.use_dp or self.trainer.use_ddp2:
+            loss_val = loss_val.unsqueeze(0)
+            acc1 = acc1.unsqueeze(0)
+            acc5 = acc5.unsqueeze(0)
+
+        output = OrderedDict({
+            'val_loss': loss_val,
+            'val_acc1': acc1,
+            'val_acc5': acc5,
+        })
+
+        return output
+
+    def validation_end(self, outputs):
+
+        val_loss_mean = 0
+        val_acc1_mean = 0
+        val_acc5_mean = 0
+
+        for output in outputs:
+            val_loss = output['val_loss']
+
+            # reduce manually when using dp
+            if self.trainer.use_dp or self.trainer.use_ddp2:
+                val_loss = torch.mean(val_loss)
+            val_loss_mean += val_loss
+
+            # reduce manually when using dp
+            val_acc1 = output['val_acc1']
+            val_acc5 = output['val_acc5']
+            if self.trainer.use_dp or self.trainer.use_ddp2:
+                val_acc1 = torch.mean(val_acc1)
+                val_acc5 = torch.mean(val_acc5)
+            val_acc1_mean += val_acc1
+            val_acc5_mean += val_acc5
+
+        val_loss_mean /= len(outputs)
+        val_acc1_mean /= len(outputs)
+        val_acc5_mean /= len(outputs)
+        tqdm_dict = {'val_loss': val_loss_mean, 'val_acc1': val_acc1_mean, 'val_acc5': val_acc5_mean}
+        result = {'progress_bar': tqdm_dict, 'log': tqdm_dict, 'val_loss': val_loss_mean}
+        return result
+
+    @classmethod
+    def __accuracy(cls, output, target, topk=(1,)):
+        """Computes the accuracy over the k top predictions for the specified values of k"""
+        with torch.no_grad():
+            maxk = max(topk)
+            batch_size = target.size(0)
+
+            _, pred = output.topk(maxk, 1, True, True)
+            pred = pred.t()
+            correct = pred.eq(target.view(1, -1).expand_as(pred))
+
+            res = []
+            for k in topk:
+                correct_k = correct[:k].view(-1).float().sum(0, keepdim=True)
+                res.append(correct_k.mul_(100.0 / batch_size))
+            return res
+
+    def configure_optimizers(self):
+        optimizer = optim.SGD(
+            self.parameters(),
+            lr=self.hparams.lr,
+            momentum=self.hparams.momentum,
+            weight_decay=self.hparams.weight_decay
+        )
+        scheduler = lr_scheduler.ExponentialLR(optimizer, gamma=0.1)
+        return [optimizer], [scheduler]
+
+    @pl.data_loader
+    def train_dataloader(self):
+        train_dir = os.path.join(self.hparams.data, 'train')
```

Comment:
is it always "train" sub-folder or shall it be a parameter (with default)?

Reply:
I think the folders being named train/val is the standard convention. I don't feel strongly either way, but I think having three arguments for data_path/train_data_path (optional)/val_data_path (optional) might be excessive.